relay-cardinality/src/limiter.rs

```diff
@@ -0,0 +1,169 @@
+    status: EntryStatus,
+}
+
+/// Represents a unique Id for a bucket within one invocation
+/// of the cardinality limiter.
+///
+/// Opaque data structure used by [`CardinalityLimiter`] to track
+/// which buckets have been accepted and rejected.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct EntryId(pub(crate) usize);
+
+/// Status wether an entry/bucket is accepted or rejected by the cardinality limiter.
+#[derive(Clone, Copy, Debug)]
+enum EntryStatus {
+    /// Entry is rejected.
+    Rejected,
+    /// Entry is accepted.
+    Accepted,
+}
+
+impl Entry {
+    /// Creates a new entry.
+    pub fn new(id: EntryId, namespace: MetricNamespace, hash: u32) -> Self {
+        Self {
+            id,
+            namespace,
+            hash,
+            status: EntryStatus::Rejected,
+        }
+    }
+
+    /// Marks the entry as accepted.
+    ///
+    /// Implementation detail, not exposed outside of the crate.
+    pub(crate) fn accept(&mut self) {
+        self.status = EntryStatus::Accepted;
+    }
+
+    /// Marks the entry as rejected.
+    ///
+    /// Implementation detail, not exposed outside of the crate.
+    pub(crate) fn reject(&mut self) {
+        self.status = EntryStatus::Rejected;
+    }
+
+    /// Whether the entry is accepted.
+    ///
+    /// Implementation detail, not exposed outside of the crate.
+    pub(crate) fn is_accepted(&self) -> bool {
+        matches!(self.status, EntryStatus::Accepted)
+    }
+}
+
+/// CardinalityLimiter configuration.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The cardinality limit to enforce.
+    pub cardinality_limit: usize,
+}
+
+/// Cardinality Limiter enforcing cardinality limits on buckets.
+///
+/// Delegates enforcement to a [`Limiter`].
+pub struct CardinalityLimiter<T: Limiter> {
+    limiter: T,
+    config: Config,
+}
+
+impl<T: Limiter> CardinalityLimiter<T> {
+    pub fn new(limiter: T, config: Config) -> Self {
+        Self { limiter, config }
+    }
+
+    /// Checks cardinality limits of a list of buckets.
+    ///
+    /// Returns an iterator of all buckets that have been accepted.
+    pub fn check_cardinality_limits(
+        &self,
+        organization: OrganizationId,
+        mut buckets: Vec<Bucket>,
+    ) -> Result<impl Iterator<Item = Bucket>> {
+        let entries = buckets
+            .iter()
+            .enumerate()
+            .filter_map(|(id, bucket)| self.parse_bucket(EntryId(id), bucket));
+
+        const EMPTY: Bucket = Bucket {
+            name: String::new(),
+            timestamp: UnixTimestamp::from_secs(0),
+            tags: BTreeMap::new(),
+            value: relay_metrics::BucketValue::Counter(0.0),
+            width: 0,
+        };
+
+        let accepted = self
+            .limiter
+            .check_cardinality_limits(organization, entries, self.config.cardinality_limit)?
+            .map(move |entry| std::mem::replace(&mut buckets[entry.id.0], EMPTY));
```

Comment:
The `swap` is a late change to be able to return owned buckets, initially the limiter took buckets by reference, but this allows us to just pass the iterator on into the partioning. 

This is a good sign that the API is not what it should be, but I think for now this is okay, I would like to use this to gather performance information and then re-work this when we come up with a plan on how to proceed with Stage 2.

Reply:
Addressed now by tracking rejections in `BTreeSet` and an iterator which can use it to skip over rejections